File: ios/Parsers/KMLParser.swift
import Foundation

final class KMLParser: NSObject, XMLParserDelegate {

  private var features: [[String: Any]] = []
  private var documentName = ""
  private var documentDescription = ""
  private var styles: [String: StyleInfo] = [:]
  private var styleMaps: [String: String] = [:]

  private var elementStack: [String] = []
  private var textBuffer = ""

  private var documentDepth = 0
  private var documentMetaCaptured = false

  private var inPlacemark = false
  private var currentFeatureId: String?
  private var currentFeatureName = ""
  private var currentFeatureDescription = ""
  private var currentFeatureStyleUrl = ""
  private var currentExtendedData: [String: Any] = [:]
  private var currentExtendedDataName = ""
  private var inExtendedData = false
  private var currentGeometryType: String?

  private var pointCoord: [Double] = []
  private var lineCoords: [[Double]] = []
  private var currentRing: [[Double]] = []
  private var outerRing: [[Double]] = []
  private var innerRings: [[[Double]]] = []
  private var inOuterBoundary = false
  private var inInnerBoundary = false

  private var inMultiGeometry = false
  private var multiGeometries: [[String: Any]] = []

  private var currentStyleId: String?
  private var buildingStyle = StyleInfo()
  private var inLineStyle = false
  private var inPolyStyle = false
  private var inIconStyle = false
  private var inIconHref = false

  private var inStyleMap = false
  private var currentStyleMapId: String?
  private var inPair = false
  private var currentPairKey = ""
  private var currentPairStyleUrl = ""

  static func parse(data: Data, sourceType: String) throws -> [String: Any] {
    let delegate = KMLParser()
    let xmlParser = XMLParser(data: data)
    xmlParser.delegate = delegate
    xmlParser.shouldProcessNamespaces = true
    xmlParser.shouldReportNamespacePrefixes = false

    guard xmlParser.parse() else {
      throw GeoParserError.parseError(xmlParser.parserError?.localizedDescription ?? "XML parse error")
    }

    var result: [String: Any] = [
      "type": "FeatureCollection",
      "sourceType": sourceType,
      "features": delegate.features
    ]
    if !delegate.documentName.isEmpty        { result["name"]        = delegate.documentName }
    if !delegate.documentDescription.isEmpty { result["description"] = delegate.documentDescription }
    return result
  }

  func parser(
    _ parser: XMLParser,
    didStartElement elementName: String,
    namespaceURI: String?,
    qualifiedName _: String?,
    attributes attributeDict: [String: String]
  ) {
    let name = stripped(elementName)
    elementStack.append(name)
    textBuffer = ""

    switch name {
    case "Document":        documentDepth += 1
    case "Style":           currentStyleId = attributeDict["id"]; buildingStyle = StyleInfo()
    case "StyleMap":        inStyleMap = true; currentStyleMapId = attributeDict["id"]
    case "Pair":            if inStyleMap { inPair = true; currentPairKey = ""; currentPairStyleUrl = "" }
    case "LineStyle":       inLineStyle = true
    case "PolyStyle":       inPolyStyle = true
    case "IconStyle":       inIconStyle = true
    case "Icon":            if inIconStyle { inIconHref = true }
    case "Placemark":
      inPlacemark = true
      currentFeatureId = attributeDict["id"]
      currentFeatureName = ""
      currentFeatureDescription = ""
      currentFeatureStyleUrl = ""
      currentExtendedData = [:]
      currentExtendedDataName = ""
      inExtendedData = false
      currentGeometryType = nil
      multiGeometries = []
    case "ExtendedData":
      if inPlacemark { inExtendedData = true }
    case "Data", "SimpleData":
      if inPlacemark && inExtendedData { currentExtendedDataName = attributeDict["name"] ?? "" }
    case "Point":           currentGeometryType = "Point";      pointCoord = []
    case "LineString":      currentGeometryType = "LineString"; lineCoords = []
    case "LinearRing":      currentRing = []
    case "Polygon":
      currentGeometryType = "Polygon"
      outerRing = []; innerRings = []
      inOuterBoundary = false; inInnerBoundary = false
    case "MultiGeometry":   inMultiGeometry = true; multiGeometries = []
    case "outerBoundaryIs": inOuterBoundary = true;  inInnerBoundary = false
    case "innerBoundaryIs": inInnerBoundary = true; inOuterBoundary = false
    default: break
    }
  }

  func parser(_ parser: XMLParser, foundCharacters string: String) {
    textBuffer += string
  }

  func parser(_ parser: XMLParser, foundCDATA block: Data) {
    if let s = String(data: block, encoding: .utf8) { textBuffer += s }
  }

  func parser(
    _ parser: XMLParser,
    didEndElement elementName: String,
    namespaceURI: String?,
    qualifiedName _: String?
  ) {
    let name = stripped(elementName)
    let text = textBuffer.trimmingCharacters(in: .whitespacesAndNewlines)
    let parent = elementStack.dropLast().last ?? ""
    defer { textBuffer = ""; if !elementStack.isEmpty { elementStack.removeLast() } }

    switch name {
    case "Document":    documentDepth -= 1
    case "name":
      if inPlacemark && parent == "Placemark" { currentFeatureName = text }
      else if isCollectionContainer(parent) && !documentMetaCaptured { documentName = text; documentMetaCaptured = true }
    case "description":
      if inPlacemark && parent == "Placemark" { currentFeatureDescription = text }
      else if isCollectionContainer(parent) && documentDescription.isEmpty { documentDescription = text }
    case "ExtendedData":
      inExtendedData = false
    case "Data":
      currentExtendedDataName = ""
    case "value":
      if inPlacemark && inExtendedData && !currentExtendedDataName.isEmpty {
        currentExtendedData[currentExtendedDataName] = parsePropertyValue(text)
      }
    case "SimpleData":
      if inPlacemark && inExtendedData && !currentExtendedDataName.isEmpty {
        currentExtendedData[currentExtendedDataName] = parsePropertyValue(text)
      }
      currentExtendedDataName = ""
    case "Style":
      if let id = currentStyleId { styles[id] = buildingStyle }
      currentStyleId = nil
    case "StyleMap":    inStyleMap = false; currentStyleMapId = nil
    case "Pair":
      if inStyleMap && currentPairKey == "normal", let mapId = currentStyleMapId {
        styleMaps[mapId] = currentPairStyleUrl
      }
      inPair = false
    case "key":         if inStyleMap && inPair { currentPairKey = text }
    case "styleUrl":
      if inPlacemark { currentFeatureStyleUrl = text }
      else if inStyleMap && inPair { currentPairStyleUrl = text }
    case "LineStyle":   inLineStyle = false
    case "PolyStyle":   inPolyStyle = false
    case "IconStyle":   inIconStyle = false
    case "Icon":        inIconHref = false
    case "color":
      let hex = kmlColorToCSS(text)
      if inLineStyle { buildingStyle.strokeColor = hex }
      else if inPolyStyle { buildingStyle.fillColor = hex }
    case "width":       if inLineStyle, let w = Double(text) { buildingStyle.strokeWidth = w }
    case "fill":        if inPolyStyle { buildingStyle.fillEnabled = (text != "0") }
    case "href":        if inIconHref { buildingStyle.iconUrl = text }
    case "scale":       if inIconStyle, let s = Double(text) { buildingStyle.iconScale = s }
    case "coordinates":
      let coords = parseCoordinates(text)
      switch parent {
      case "Point":      pointCoord = coords.first ?? []
      case "LineString": lineCoords = coords
      case "LinearRing": currentRing = coords
      default: break
      }
    case "LinearRing":
      if inOuterBoundary { outerRing = currentRing }
      else if inInnerBoundary { innerRings.append(currentRing) }
    case "outerBoundaryIs": inOuterBoundary = false
    case "innerBoundaryIs": inInnerBoundary = false
    case "Point", "LineString", "Polygon":
      if inMultiGeometry, let geom = buildGeometry(type: name) {
        multiGeometries.append(geom)
        pointCoord = []; lineCoords = []; outerRing = []; innerRings = []
      }
    case "MultiGeometry":   inMultiGeometry = false; currentGeometryType = "MultiGeometry"
    case "Placemark":       finalizeFeature(); inPlacemark = false
    default: break
    }
  }

  private func stripped(_ name: String) -> String {
    if let r = name.range(of: ":", options: .backwards) { return String(name[r.upperBound...]) }
    return name
  }

  private func isCollectionContainer(_ name: String) -> Bool {
    name == "Document" || name == "Folder"
  }

  private func kmlColorToCSS(_ kml: String) -> String {
    let s = kml.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard s.count == 8 else { return "#000000" }
    return "#\(s.dropFirst(6).prefix(2))\(s.dropFirst(4).prefix(2))\(s.dropFirst(2).prefix(2))"
  }

  private func parseCoordinates(_ text: String) -> [[Double]] {
    text
      .components(separatedBy: .whitespacesAndNewlines)
      .filter { !$0.isEmpty }
      .compactMap { tuple -> [Double]? in
        let parts = tuple.split(separator: ",").compactMap { Double($0) }
        return parts.count >= 2 ? Array(parts) : nil
      }
  }

  private func parsePropertyValue(_ text: String) -> Any {
    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    let lowercased = trimmed.lowercased()

    if lowercased == "true" { return true }
    if lowercased == "false" { return false }
    if isNumericScalar(trimmed), let number = Double(trimmed) {
      return number
    }

    return trimmed
  }

  private func isNumericScalar(_ text: String) -> Bool {
    guard !text.isEmpty else { return false }
    return text.range(of: #"^-?(0|[1-9]\d*)(\.\d+)?$"#, options: .regularExpression) != nil
  }

  private func buildGeometry(type: String) -> [String: Any]? {
    switch type {
    case "Point":
      guard !pointCoord.isEmpty else { return nil }
      return ["type": "Point", "coordinates": pointCoord]
    case "LineString":
      guard !lineCoords.isEmpty else { return nil }
      return ["type": "LineString", "coordinates": lineCoords]
    case "Polygon":
      guard !outerRing.isEmpty else { return nil }
      return ["type": "Polygon", "coordinates": [outerRing] + innerRings]
    default:
      return nil
    }
  }

  private func resolveStyle(url: String) -> StyleInfo? {
    let id = url.hasPrefix("#") ? String(url.dropFirst()) : url
    if let style = styles[id] { return style }
    if let normalUrl = styleMaps[id] {
      let nid = normalUrl.hasPrefix("#") ? String(normalUrl.dropFirst()) : normalUrl
      return styles[nid]
    }
    return nil
  }

  private func finalizeFeature() {
    var properties: [String: Any] = [:]
    if !currentFeatureName.isEmpty { properties["name"] = currentFeatureName }
    for (key, value) in currentExtendedData {
      properties[key] = value
    }

    if !currentFeatureDescription.isEmpty {
      let extracted = extractDescriptionAttributes(from: currentFeatureDescription)
      for (key, value) in extracted.attributes where properties[key] == nil {
        properties[key] = value
      }
      if let description = extracted.description {
        properties["description"] = description
      }
    }

    if !currentFeatureStyleUrl.isEmpty {
      properties["styleId"] = currentFeatureStyleUrl
      if let style = resolveStyle(url: currentFeatureStyleUrl), !style.isEmpty {
        properties["style"] = style.asDictionary()
      }
    }

    let geometries: [[String: Any]]
    if currentGeometryType == "MultiGeometry" {
      guard !multiGeometries.isEmpty else { return }
      geometries = multiGeometries
    } else if let gType = currentGeometryType, let geom = buildGeometry(type: gType) {
      geometries = [geom]
    } else {
      return
    }

    for (index, geometry) in geometries.enumerated() {
      var feature: [String: Any] = ["type": "Feature", "geometry": geometry, "properties": properties]
      if let id = splitFeatureId(baseId: currentFeatureId, index: index, total: geometries.count) {
        feature["id"] = id
      }
      features.append(feature)
    }
  }

  private func splitFeatureId(baseId: String?, index: Int, total: Int) -> String? {
    guard let baseId, !baseId.isEmpty else { return nil }
    guard total > 1 else { return baseId }
    return "\(baseId)_\(index + 1)"
  }

  private func extractDescriptionAttributes(from description: String) -> (attributes: [String: Any], description: String?) {
    let textDescription = stripHTML(description)
    guard description.contains("<table"), description.contains("<th"), description.contains("<td") else {
      return ([:], textDescription.isEmpty ? nil : textDescription)
    }

    guard let regex = try? NSRegularExpression(
      pattern: #"<th[^>]*>\s*(.*?)\s*</th>\s*<td[^>]*>\s*(.*?)\s*</td>"#,
      options: [.caseInsensitive, .dotMatchesLineSeparators]
    ) else {
      return ([:], textDescription.isEmpty ? nil : textDescription)
    }

    let nsDescription = description as NSString
    let matches = regex.matches(in: description, range: NSRange(location: 0, length: nsDescription.length))

    var attributes: [String: Any] = [:]
    for match in matches where match.numberOfRanges >= 3 {
      let key = stripHTML(nsDescription.substring(with: match.range(at: 1)))
      if key.isEmpty { continue }
      let value = stripHTML(nsDescription.substring(with: match.range(at: 2)))
      attributes[key] = parsePropertyValue(value)
    }

    let cleanedDescription: String?
    if !textDescription.isEmpty && !(matches.isEmpty == false && textDescription == "Attributes") {
      cleanedDescription = textDescription
    } else {
      cleanedDescription = nil
    }

    return (attributes, cleanedDescription)
  }

  private func stripHTML(_ value: String) -> String {
    value
      .replacingOccurrences(of: #"<[^>]+>"#, with: " ", options: .regularExpression)
      .replacingOccurrences(of: "&nbsp;", with: " ")
      .replacingOccurrences(of: "&amp;", with: "&")
      .replacingOccurrences(of: "&lt;", with: "<")
      .replacingOccurrences(of: "&gt;", with: ">")
      .replacingOccurrences(of: "&quot;", with: "\"")
      .replacingOccurrences(of: "&#39;", with: "'")
      .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)
  }
}

private struct StyleInfo {
  var strokeColor: String?
  var strokeWidth: Double?
  var fillColor: String?
  var fillEnabled: Bool = true
  var iconUrl: String?
  var iconScale: Double?

  var isEmpty: Bool { strokeColor == nil && fillColor == nil && iconUrl == nil }

  func asDictionary() -> [String: Any] {
    var d: [String: Any] = [:]
    if let v = strokeColor { d["strokeColor"] = v }
    if let v = strokeWidth { d["strokeWidth"] = v }
    if let v = fillColor   { d["fillColor"]   = v }
    if !fillEnabled        { d["fillEnabled"] = false }
    if let v = iconUrl     { d["iconUrl"]     = v }
    if let v = iconScale   { d["iconScale"]   = v }
    return d
  }
}

File: android/src/main/java/expo/modules/geoparser/parsers/KMLParser.kt
package expo.modules.geoparser.parsers

import expo.modules.geoparser.GeoParserError
import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler
import java.io.InputStream
import javax.xml.parsers.SAXParserFactory

class KMLParser(
    private val batchSize: Int = 200,
    private val onFeatures: (List<Map<String, Any?>>, isLast: Boolean) -> Unit
) : DefaultHandler() {

    private val featureBuffer = mutableListOf<Map<String, Any?>>()
    var documentName = ""
    var documentDescription = ""
    private val styles = mutableMapOf<String, StyleInfo>()
    private val styleMaps = mutableMapOf<String, String>()

    private val elementStack = mutableListOf<String>()
    private val textBuffer = StringBuilder()

    private var documentDepth = 0
    private var documentMetaCaptured = false

    private var inPlacemark = false
    private var currentFeatureId: String? = null
    private var currentFeatureName = ""
    private var currentFeatureDescription = ""
    private var currentFeatureStyleUrl = ""
    private var currentExtendedData = linkedMapOf<String, Any?>()
    private var currentExtendedDataName = ""
    private var inExtendedData = false
    private var currentGeometryType: String? = null

    private var pointCoord = listOf<Double>()
    private var lineCoords = listOf<List<Double>>()
    private var currentRing = listOf<List<Double>>()
    private var outerRing = listOf<List<Double>>()
    private var innerRings = mutableListOf<List<List<Double>>>()
    private var inOuterBoundary = false
    private var inInnerBoundary = false

    private var inMultiGeometry = false
    private var multiGeometries = mutableListOf<Map<String, Any?>>()

    private var currentStyleId: String? = null
    private var buildingStyle = StyleInfo()
    private var inLineStyle = false
    private var inPolyStyle = false
    private var inIconStyle = false
    private var inIconHref = false

    private var inStyleMap = false
    private var currentStyleMapId: String? = null
    private var inPair = false
    private var currentPairKey = ""
    private var currentPairStyleUrl = ""

    companion object {
        fun parse(input: InputStream, sourceType: String, onFeatures: (List<Map<String, Any?>>, Boolean) -> Unit): Map<String, Any?> {
            val handler = KMLParser(onFeatures = onFeatures)
            val factory = SAXParserFactory.newInstance().apply { isNamespaceAware = true }
            try {
                factory.newSAXParser().parse(input, handler)
            } catch (e: Exception) {
                throw GeoParserError.ParseError(e.message ?: "XML parse error")
            }
            val result = mutableMapOf<String, Any?>(
                "type" to "FeatureCollection",
                "sourceType" to sourceType
            )
            if (handler.documentName.isNotEmpty()) result["name"] = handler.documentName
            if (handler.documentDescription.isNotEmpty()) result["description"] = handler.documentDescription
            return result
        }
    }

    override fun startElement(uri: String?, localName: String?, qName: String?, attrs: Attributes?) {
        val name = stripped(localName ?: qName ?: "")
        elementStack.add(name)
        textBuffer.clear()

        when (name) {
            "Document"       -> documentDepth++
            "Style"          -> { currentStyleId = attrs?.getValue("id"); buildingStyle = StyleInfo() }
            "StyleMap"       -> { inStyleMap = true; currentStyleMapId = attrs?.getValue("id") }
            "Pair"           -> if (inStyleMap) { inPair = true; currentPairKey = ""; currentPairStyleUrl = "" }
            "LineStyle"      -> inLineStyle = true
            "PolyStyle"      -> inPolyStyle = true
            "IconStyle"      -> inIconStyle = true
            "Icon"           -> if (inIconStyle) inIconHref = true
            "Placemark"      -> {
                inPlacemark = true
                currentFeatureId = attrs?.getValue("id")
                currentFeatureName = ""
                currentFeatureDescription = ""
                currentFeatureStyleUrl = ""
                currentExtendedData = linkedMapOf()
                currentExtendedDataName = ""
                inExtendedData = false
                currentGeometryType = null
                multiGeometries = mutableListOf()
            }
            "ExtendedData"   -> if (inPlacemark) inExtendedData = true
            "Data", "SimpleData" -> if (inPlacemark && inExtendedData) currentExtendedDataName = attrs?.getValue("name") ?: ""
            "Point"          -> { currentGeometryType = "Point";      pointCoord = listOf() }
            "LineString"     -> { currentGeometryType = "LineString"; lineCoords = listOf() }
            "LinearRing"     -> currentRing = listOf()
            "Polygon"        -> {
                currentGeometryType = "Polygon"
                outerRing = listOf(); innerRings = mutableListOf()
                inOuterBoundary = false; inInnerBoundary = false
            }
            "MultiGeometry"  -> { inMultiGeometry = true; multiGeometries = mutableListOf() }
            "outerBoundaryIs" -> { inOuterBoundary = true;  inInnerBoundary = false }
            "innerBoundaryIs" -> { inInnerBoundary = true; inOuterBoundary = false }
        }
    }

    override fun characters(ch: CharArray?, start: Int, length: Int) {
        if (ch != null) textBuffer.append(ch, start, length)
    }

    override fun endElement(uri: String?, localName: String?, qName: String?) {
        val name = stripped(localName ?: qName ?: "")
        val text = textBuffer.toString().trim()
        val parent = elementStack.dropLast(1).lastOrNull() ?: ""
        textBuffer.clear()
        if (elementStack.isNotEmpty()) elementStack.removeAt(elementStack.lastIndex)

        when (name) {
            "Document"   -> documentDepth--
            "name"       -> when {
                inPlacemark && parent == "Placemark" -> currentFeatureName = text
                isCollectionContainer(parent) && !documentMetaCaptured -> {
                    documentName = text
                    documentMetaCaptured = true
                }
            }
            "description" -> when {
                inPlacemark && parent == "Placemark" -> currentFeatureDescription = text
                isCollectionContainer(parent) && documentDescription.isEmpty() -> documentDescription = text
            }
            "ExtendedData" -> inExtendedData = false
            "Data" -> currentExtendedDataName = ""
            "value" -> if (inPlacemark && inExtendedData && currentExtendedDataName.isNotEmpty()) {
                currentExtendedData[currentExtendedDataName] = parsePropertyValue(text)
            }
            "SimpleData" -> {
                if (inPlacemark && inExtendedData && currentExtendedDataName.isNotEmpty()) {
                    currentExtendedData[currentExtendedDataName] = parsePropertyValue(text)
                }
                currentExtendedDataName = ""
            }
            "Style"      -> { currentStyleId?.let { styles[it] = buildingStyle }; currentStyleId = null }
            "StyleMap"   -> { inStyleMap = false; currentStyleMapId = null }
            "Pair"       -> {
                if (inStyleMap && currentPairKey == "normal") currentStyleMapId?.let { styleMaps[it] = currentPairStyleUrl }
                inPair = false
            }
            "key"        -> if (inStyleMap && inPair) currentPairKey = text
            "styleUrl"   -> when {
                inPlacemark -> currentFeatureStyleUrl = text
                inStyleMap && inPair -> currentPairStyleUrl = text
            }
            "LineStyle"  -> inLineStyle = false
            "PolyStyle"  -> inPolyStyle = false
            "IconStyle"  -> inIconStyle = false
            "Icon"       -> inIconHref = false
            "color"      -> {
                val hex = kmlColorToCSS(text)
                when {
                    inLineStyle -> buildingStyle.strokeColor = hex
                    inPolyStyle -> buildingStyle.fillColor = hex
                }
            }
            "width"      -> if (inLineStyle) text.toDoubleOrNull()?.let { buildingStyle.strokeWidth = it }
            "fill"       -> if (inPolyStyle) buildingStyle.fillEnabled = text != "0"
            "href"       -> if (inIconHref) buildingStyle.iconUrl = text
            "scale"      -> if (inIconStyle) text.toDoubleOrNull()?.let { buildingStyle.iconScale = it }
            "coordinates" -> {
                val coords = parseCoordinates(text)
                when (parent) {
                    "Point"      -> pointCoord = coords.firstOrNull() ?: listOf()
                    "LineString" -> lineCoords = coords
                    "LinearRing" -> currentRing = coords
                }
            }
            "LinearRing"      -> when {
                inOuterBoundary -> outerRing = currentRing
                inInnerBoundary -> innerRings.add(currentRing)
            }
            "outerBoundaryIs" -> inOuterBoundary = false
            "innerBoundaryIs" -> inInnerBoundary = false
            "Point", "LineString", "Polygon" -> {
                if (inMultiGeometry) {
                    buildGeometry(name)?.let {
                        multiGeometries.add(it)
                        pointCoord = listOf(); lineCoords = listOf()
                        outerRing = listOf(); innerRings = mutableListOf()
                    }
                }
            }
            "MultiGeometry" -> { inMultiGeometry = false; currentGeometryType = "MultiGeometry" }
            "Placemark"     -> { finalizeFeature(); inPlacemark = false }
        }
    }

    override fun endDocument() {
        onFeatures(featureBuffer.toList(), true)
        featureBuffer.clear()
    }

    private fun stripped(name: String): String {
        val idx = name.lastIndexOf(':')
        return if (idx >= 0) name.substring(idx + 1) else name
    }

    private fun isCollectionContainer(name: String): Boolean =
        name == "Document" || name == "Folder"

    private fun kmlColorToCSS(kml: String): String {
        val s = kml.trim().lowercase()
        if (s.length != 8) return "#000000"
        return "#${s.substring(6, 8)}${s.substring(4, 6)}${s.substring(2, 4)}"
    }

    private fun parseCoordinates(text: String): List<List<Double>> =
        text.trim().split(Regex("\\s+"))
            .filter { it.isNotEmpty() }
            .mapNotNull { tuple ->
                val parts = tuple.split(",").mapNotNull { it.toDoubleOrNull() }
                if (parts.size >= 2) parts else null
            }

    private fun parsePropertyValue(text: String): Any {
        val trimmed = text.trim()
        return when {
            trimmed.equals("true", ignoreCase = true) -> true
            trimmed.equals("false", ignoreCase = true) -> false
            NUMERIC_SCALAR.matches(trimmed) -> trimmed.toDouble()
            else -> trimmed
        }
    }

    private fun buildGeometry(type: String): Map<String, Any?>? = when (type) {
        "Point"      -> if (pointCoord.isNotEmpty()) mapOf("type" to "Point", "coordinates" to pointCoord) else null
        "LineString" -> if (lineCoords.isNotEmpty()) mapOf("type" to "LineString", "coordinates" to lineCoords) else null
        "Polygon"    -> if (outerRing.isNotEmpty()) mapOf("type" to "Polygon", "coordinates" to (listOf(outerRing) + innerRings)) else null
        else         -> null
    }

    private fun resolveStyle(url: String): StyleInfo? {
        val id = if (url.startsWith("#")) url.drop(1) else url
        styles[id]?.let { return it }
        styleMaps[id]?.let { normalUrl ->
            val nid = if (normalUrl.startsWith("#")) normalUrl.drop(1) else normalUrl
            return styles[nid]
        }
        return null
    }

    private fun finalizeFeature() {
        val properties = mutableMapOf<String, Any?>()
        if (currentFeatureName.isNotEmpty()) properties["name"] = currentFeatureName
        properties.putAll(currentExtendedData)
        if (currentFeatureDescription.isNotEmpty()) {
            val extracted = extractDescriptionAttributes(currentFeatureDescription)
            extracted.attributes.forEach { (key, value) -> properties.putIfAbsent(key, value) }
            extracted.description?.let { properties["description"] = it }
        }
        if (currentFeatureStyleUrl.isNotEmpty()) {
            properties["styleId"] = currentFeatureStyleUrl
            resolveStyle(currentFeatureStyleUrl)?.takeIf { !it.isEmpty }?.let {
                properties["style"] = it.asDictionary()
            }
        }

        val geometries: List<Map<String, Any?>> = if (currentGeometryType == "MultiGeometry") {
            if (multiGeometries.isEmpty()) return
            multiGeometries.toList()
        } else {
            val gType = currentGeometryType ?: return
            listOf(buildGeometry(gType) ?: return)
        }

        geometries.forEachIndexed { index, geometry ->
            val feature = mutableMapOf<String, Any?>(
                "type" to "Feature",
                "geometry" to geometry,
                "properties" to properties
            )
            splitFeatureId(currentFeatureId, index, geometries.size)?.let { feature["id"] = it }

            featureBuffer.add(feature)
            if (featureBuffer.size >= batchSize) {
                onFeatures(featureBuffer.toList(), false)
                featureBuffer.clear()
            }
        }
    }

    private fun splitFeatureId(baseId: String?, index: Int, total: Int): String? {
        if (baseId.isNullOrEmpty()) return null
        if (total <= 1) return baseId
        return "${baseId}_${index + 1}"
    }

    private fun extractDescriptionAttributes(description: String): ExtractedDescription {
        val textDescription = stripHtml(description)
        if (!description.contains("<table") || !description.contains("<th") || !description.contains("<td")) {
            return ExtractedDescription(linkedMapOf(), textDescription.ifEmpty { null })
        }

        val attributes = linkedMapOf<String, Any?>()
        for (match in HTML_ATTRIBUTE_ROW.findAll(description)) {
            val key = stripHtml(match.groupValues[1])
            if (key.isEmpty()) continue
            val value = stripHtml(match.groupValues[2])
            attributes[key] = parsePropertyValue(value)
        }

        val cleanedDescription =
            if (textDescription.isNotEmpty() && !(attributes.isNotEmpty() && textDescription == "Attributes")) {
                textDescription
            } else {
                null
            }

        return ExtractedDescription(attributes, cleanedDescription)
    }

    private fun stripHtml(value: String): String =
        value
            .replace(HTML_TAG, " ")
            .replace("&nbsp;", " ")
            .replace("&amp;", "&")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .replace(WHITESPACE, " ")
            .trim()

    private data class ExtractedDescription(
        val attributes: LinkedHashMap<String, Any?>,
        val description: String?
    )
}

private data class StyleInfo(
    var strokeColor: String? = null,
    var strokeWidth: Double? = null,
    var fillColor: String? = null,
    var fillEnabled: Boolean = true,
    var iconUrl: String? = null,
    var iconScale: Double? = null
) {
    val isEmpty: Boolean get() = strokeColor == null && fillColor == null && iconUrl == null

    fun asDictionary(): Map<String, Any?> {
        val d = mutableMapOf<String, Any?>()
        strokeColor?.let { d["strokeColor"] = it }
        strokeWidth?.let { d["strokeWidth"] = it }
        fillColor?.let { d["fillColor"] = it }
        if (!fillEnabled) d["fillEnabled"] = false
        iconUrl?.let { d["iconUrl"] = it }
        iconScale?.let { d["iconScale"] = it }
        return d
    }
}

private val HTML_TAG = Regex("<[^>]+>")
private val HTML_ATTRIBUTE_ROW = Regex(
    "<th[^>]*>\\s*(.*?)\\s*</th>\\s*<td[^>]*>\\s*(.*?)\\s*</td>",
    setOf(RegexOption.IGNORE_CASE, RegexOption.DOT_MATCHES_ALL)
)
private val WHITESPACE = Regex("\\s+")
private val NUMERIC_SCALAR = Regex("^-?(0|[1-9]\\d*)(\\.\\d+)?$")
